File: app/src/main/java/com/commonpepper/photosen/network/DownloadService.java
package com.commonpepper.photosen.network;

import android.app.IntentService;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.WallpaperManager;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;

import com.commonpepper.photosen.Photosen;
import com.commonpepper.photosen.R;
import com.commonpepper.photosen.ui.activities.CropActivity;
import com.google.firebase.analytics.FirebaseAnalytics;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;
import androidx.core.content.FileProvider;

public class DownloadService extends IntentService {
    private static final String TAG = DownloadService.class.getSimpleName();

    public static boolean isRunning = false;

    private NotificationManager mgr;

    public enum Aciton {
        DOWNLOAD_ONLY,
        WALLPAPER
    }

    public static final String TAG_URL = "url";
    public static final String TAG_FILENAME = "filename";
    public static final String TAG_ACTION = "action";

    private static final String CHANNEL_ID = "download_channel_id";
    public static final int FOREGROUND_ID = 98732517;
    public static final int NOTIFY_ID = 98732518;

    private String filename;
    private Aciton aciton;

    public DownloadService() {
        super("PhotosenDownloadService");
    }

    @Override
    public void onCreate() {
        isRunning = true;
        super.onCreate();
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        Photosen.getFirebaseAnalytics().logEvent("Download_start", null);
        String urlToDownload = intent.getStringExtra(TAG_URL);

        filename = intent.getStringExtra(TAG_FILENAME);
        aciton = (Aciton) intent.getSerializableExtra(TAG_ACTION);

        mgr = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && mgr.getNotificationChannel(CHANNEL_ID) == null) {
            NotificationChannel c = new NotificationChannel(CHANNEL_ID, "Photosen", NotificationManager.IMPORTANCE_DEFAULT);

            c.setSound(null, null);
            c.enableVibration(false);

            mgr.createNotificationChannel(c);
        }

        startForeground(FOREGROUND_ID, buildForegroundNotification());

        String path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
                .getPath() + File.separator + filename;

        try {
            File output = new File(path);

            if (output.exists()) {
                output.delete();
            }

            URL url = new URL(urlToDownload);
            HttpURLConnection c = (HttpURLConnection) url.openConnection();
            int fileLength = c.getContentLength();

            FileOutputStream fos = new FileOutputStream(output.getPath());
            BufferedOutputStream out = new BufferedOutputStream(fos);

            try {
                InputStream in = c.getInputStream();
                byte[] buffer = new byte[8192];
                int len = 0;
                int sum = 0;

                while ((len = in.read(buffer)) >= 0) {
                    out.write(buffer, 0, len);
                    sum += len;
                    progressNotification(fileLength, sum);
                }

                out.flush();
            } finally {
                fos.getFD().sync();
                out.close();
                c.disconnect();
            }

            Uri uri = FileProvider.getUriForFile(this, Photosen.PACKAGE_NAME + ".fileprovider", new File(path));
            sendBroadcast(new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, uri));
            if (aciton == Aciton.WALLPAPER) {
                try {
                    Intent wallpaperIntent = WallpaperManager.getInstance(this).getCropAndSetWallpaperIntent(uri);
                    wallpaperIntent.setDataAndType(uri, "image/*");
                    wallpaperIntent.putExtra("mimeType", "image/*");
                    wallpaperIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    wallpaperIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                    wallpaperIntent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
                    startActivity(wallpaperIntent);
                    Photosen.getFirebaseAnalytics().logEvent("Crop_and_set_wallpaper_intent", null);
                } catch (IllegalArgumentException e) {
                    //can't crop and set with default methods
                    Intent myCrop = new Intent(this, CropActivity.class);
                    myCrop.putExtra(CropActivity.TAG_URISTR, uri.toString());
                    myCrop.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(myCrop);
                    Photosen.getFirebaseAnalytics().logEvent("My_crop", null);
                }
            }

            completeNotification(uri, null);
        } catch (IOException e) {
            completeNotification(null, e);
        } finally {
            mgr.cancel(FOREGROUND_ID); //on some devices stopForeground don't remove notification
            stopForeground(true);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        isRunning = false;
    }

    private void progressNotification(int total, int current) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, CHANNEL_ID);
        builder.setOnlyAlertOnce(true);
        builder.setContentTitle(getActionString())
                .setContentText(filename)
                .setSmallIcon(android.R.drawable.stat_sys_download)
                .setProgress(total, current, false);

        mgr.notify(FOREGROUND_ID, builder.build());
    }

    private void completeNotification(Uri uri, Exception e) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, CHANNEL_ID);
        builder.setAutoCancel(true).setOngoing(false).setWhen(System.currentTimeMillis()).setProgress(0, 0, false);

        if (e == null) {
            builder.setContentTitle(getString(R.string.download_complete))
                    .setContentText(getString(R.string.click_here_to_open) + " " + filename)
                    .setSmallIcon(android.R.drawable.stat_sys_download_done);

            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            Intent chooser = Intent.createChooser(intent, getString(R.string.open_file_with));
            intent.setDataAndType(uri, "image/*");
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
            PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, chooser, PendingIntent.FLAG_CANCEL_CURRENT);

            builder.setContentIntent(pendingIntent);
        } else {
            builder.setContentTitle(getString(R.string.download_error))
                    .setContentText(e.getMessage())
                    .setSmallIcon(android.R.drawable.stat_notify_error);
        }

        mgr.notify(NOTIFY_ID, builder.build());
    }

    private Notification buildForegroundNotification() {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, CHANNEL_ID);
        builder.setOngoing(true)
                .setContentTitle(getActionString())
                .setContentText(filename)
                .setSmallIcon(android.R.drawable.stat_sys_download);

        return builder.build();
    }

    private String getActionString() {
        return aciton == Aciton.DOWNLOAD_ONLY ? getString(R.string.downloading_image) : getString(R.string.downloading_wallpaper);
    }
}

File: app/src/main/java/com/commonpepper/photosen/network/DownloadService.kt
package com.commonpepper.photosen.network

import android.R.drawable
import android.app.*
import android.content.Context
import android.content.Intent
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import androidx.core.app.NotificationCompat.Builder
import androidx.core.content.FileProvider
import com.commonpepper.photosen.Photosen
import com.commonpepper.photosen.Photosen.Companion.firebaseAnalytics
import com.commonpepper.photosen.R.string
import com.commonpepper.photosen.ui.activities.CropActivity
import java.io.*
import java.net.HttpURLConnection
import java.net.URL


class DownloadService : IntentService("PhotosenDownloadService") {
    private var mgr: NotificationManager? = null

    enum class Aciton { DOWNLOAD_ONLY, WALLPAPER }

    private var filename: String? = null
    private var aciton: Aciton? = null
    override fun onCreate() {
        isRunning = true
        super.onCreate()
    }

    override fun onHandleIntent(intent: Intent?) {
        firebaseAnalytics.logEvent("Download_start", null)
        val urlToDownload: String? = intent!!.getStringExtra(TAG_URL)
        filename = intent.getStringExtra(TAG_FILENAME)
        aciton = intent.getSerializableExtra(TAG_ACTION) as Aciton
        mgr = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        if (VERSION.SDK_INT >= VERSION_CODES.O && mgr!!.getNotificationChannel(CHANNEL_ID) == null) {
            val c = NotificationChannel(CHANNEL_ID, "Photosen", NotificationManager.IMPORTANCE_DEFAULT)
            c.setSound(null, null)
            c.enableVibration(false)
            mgr!!.createNotificationChannel(c)
        }
        startForeground(FOREGROUND_ID, buildForegroundNotification())
        val path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
                .path + File.separator + filename
        try {
            val output = File(path)
            if (output.exists()) {
                output.delete()
            }
            val url = URL(urlToDownload)
            val c = url.openConnection() as HttpURLConnection
            val fileLength = c.contentLength
            val fos = FileOutputStream(output.path)
            val out = BufferedOutputStream(fos)
            try {
                val `in`: InputStream = c.inputStream
                val buffer = ByteArray(8192)
                var len: Int
                var sum = 0
                while (`in`.read(buffer).also { len = it } >= 0) {
                    out.write(buffer, 0, len)
                    sum += len
                    progressNotification(fileLength, sum)
                }
                out.flush()
            } finally {
                fos.fd.sync()
                out.close()
                c.disconnect()
            }
            val uri: Uri = FileProvider.getUriForFile(this, Photosen.PACKAGE_NAME + ".fileprovider", File(path))

            //Different ways of scan image work on different devices
            sendBroadcast(Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE, uri))
            MediaScannerConnection.scanFile(applicationContext, arrayOf(path), arrayOf("image/*")) { _, _->}
//            if (VERSION.SDK_INT >= VERSION_CODES.Q) {
//                MediaStore.setIncludePending(uri)
//            } else {
//                MediaStore.Images.Media.insertImage(contentResolver, path, filename, "")
//            }

            if (aciton == Aciton.WALLPAPER) {
                try {
                    val wallpaperIntent: Intent = WallpaperManager.getInstance(this).getCropAndSetWallpaperIntent(uri)
                    wallpaperIntent.setDataAndType(uri, "image/*")
                    wallpaperIntent.putExtra("mimeType", "image/*")
                    wallpaperIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    wallpaperIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                    wallpaperIntent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
                    startActivity(wallpaperIntent)
                    firebaseAnalytics.logEvent("Crop_and_set_wallpaper_intent", null)
                } catch (e: IllegalArgumentException) {
                    //can't crop and set with default methods

                    val myCrop = Intent(this, CropActivity::class.java)
                    myCrop.putExtra(CropActivity.TAG_URISTR, uri.toString())
                    myCrop.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(myCrop)
                    firebaseAnalytics.logEvent("My_crop", null)
                }
            }
            completeNotification(uri, null)
        } catch (e: IOException) {
            completeNotification(null, e)
        } finally {
            mgr!!.cancel(FOREGROUND_ID) //on some devices stopForeground don't remove notification

            stopForeground(true)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        isRunning = false
    }

    private fun progressNotification(total: Int, current: Int) {
        val builder = Builder(this, CHANNEL_ID)
        builder.setOnlyAlertOnce(true)
        builder.setContentTitle(actionString)
                .setContentText(filename)
                .setSmallIcon(drawable.stat_sys_download)
                .setProgress(total, current, false)
        mgr!!.notify(FOREGROUND_ID, builder.build())
    }

    private fun completeNotification(uri: Uri?, e: Exception?) {
        val builder = Builder(this, CHANNEL_ID)
        builder.setAutoCancel(true).setOngoing(false).setWhen(System.currentTimeMillis()).setProgress(0, 0, false)
        if (e == null) {
            builder.setContentTitle(getString(string.download_complete))
                    .setContentText(getString(string.click_here_to_open) + " " + filename)
                    .setSmallIcon(drawable.stat_sys_download_done)
            val intent = Intent(Intent.ACTION_VIEW)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            val chooser: Intent? = Intent.createChooser(intent, getString(string.open_file_with))
            intent.setDataAndType(uri, "image/*")
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
            val pendingIntent: PendingIntent? = PendingIntent.getActivity(this, 0, chooser, PendingIntent.FLAG_CANCEL_CURRENT)
            builder.setContentIntent(pendingIntent)
        } else {
            Log.d(TAG, e.message!!)
            builder.setContentTitle(getString(string.download_error))
                    .setContentText(e.message)
                    .setSmallIcon(drawable.stat_notify_error)
        }
        mgr!!.notify(NOTIFY_ID, builder.build())
    }

    private fun buildForegroundNotification(): Notification? {
        val builder = Builder(this, CHANNEL_ID)
        builder.setOngoing(true)
                .setContentTitle(actionString)
                .setContentText(filename)
                .setSmallIcon(drawable.stat_sys_download)
        return builder.build()
    }

    private val actionString: String
        get() = if (aciton == Aciton.DOWNLOAD_ONLY) getString(string.downloading_image) else getString(string.downloading_wallpaper)

    companion object {
        private val TAG = DownloadService::class.java.simpleName
        var isRunning = false
        const val TAG_URL = "url"
        const val TAG_FILENAME = "filename"
        const val TAG_ACTION = "action"
        private const val CHANNEL_ID = "download_channel_id"
        const val FOREGROUND_ID = 98732517
        const val NOTIFY_ID = 98732518
    }
}
